File: Semestr III/ANL/Lista 4/zad4.cpp
#include <iostream>
#include <iomanip>
#include <math.h>

double f(double x)
{
    return pow(x, 2) - 2.0 * cos(3.0 * x + 1.0);
}

int steps = 16;
double e = pow(10, -5);

void bisection(double a0, double b0)
{
    double a = a0;
    double b = b0;
    double middle = (a + b) / 2.0;
    bool increasing = f(b0) > 0;

    for(int i = 0; i < steps; i++)
    {
        double fm = f(middle);

        if(increasing) 
        {
            if(fm > 0) b = middle;
            else a = middle;
        }
        else
        {
            if(fm < 0) b = middle;
            else a = middle;
        }
        

        middle = (a + b) / 2.0;
    }
    std::cout << std::fixed << "x" << int(a0+1) << " = " << middle << std::endl;
}

int main()
{
    bisection(-1, 0);
    bisection(0, 1);
    return 0;
}

File: Semestr III/ANL/Lista 4/zad4.py
from math import cos

def f(x):
    return x**2 - 2 * cos(3 * x + 1)

intervals = [[-1, 0], [0, 1]]
for interval in intervals:
    a, b = interval
    increasing = f(a) < 0
    while True:
        mid = (a+b) / 2.
        val = f(mid)
        if b - a < 2 * 10**(-5):
            print(f"x0={mid}, [a, b]=[{a}, {b}]")
            break
        if increasing:
            if val > 0:
                b = mid
            else:
                a = mid
        else:
            if val > 0:
                a = mid
            else:
                b = mid